src/mpid/ch4/shm/xpmem/xpmem_control.c

```diff
@@ -129,3 +161,120 @@ int MPIDI_XPMEM_ctrl_send_lmt_req_cb(MPIDI_SHM_ctrl_hdr_t * ctrl_hdr)
+        slmt_req_hdr = &dmessage->ctrl_hdr.xpmem_slmt_req;
+        if (!OPA_load_int(&MPIDI_XPMEM_global.coop_counter
+                          [MPIDI_XPMEM_global.local_rank * MPIDI_XPMEM_global.num_local +
+                           slmt_req_hdr->src_lrank])) {
+            DL_DELETE(MPIDI_XPMEM_global.dmessage_queue, dmessage);
+            mpi_errno = MPIDI_XPMEM_ctrl_send_lmt_req_cb(&dmessage->ctrl_hdr);
+            MPIR_ERR_CHECK(mpi_errno);
+
+            MPL_free(dmessage);
+        }
+    }
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_XPMEM_CTRL_RECV_LMT_ACK_CB);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+int MPIDI_XPMEM_ctrl_send_lmt_rts_req_cb(MPIDI_SHM_ctrl_hdr_t * ctrl_hdr)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIDI_XPMEM_am_dmessage_t *dmessage;
+    MPIDI_SHM_ctrl_xpmem_send_lmt_req_t *slmt_req_hdr =
+        &((MPIDI_SHM_ctrl_hdr_t *) ctrl_hdr)->xpmem_slmt_req;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_XPMEM_CTRL_SEND_LMT_RTS_REQ_CB);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_CTRL_SEND_LMT_RTS_REQ_CB);
+
+    if (OPA_load_int(&MPIDI_XPMEM_global.coop_counter
+                     [MPIDI_XPMEM_global.local_rank * MPIDI_XPMEM_global.num_local +
+                      slmt_req_hdr->src_lrank])) {
+        /* Previous copy is not done, need to delay current copy to next round. */
+        dmessage =
+            (MPIDI_XPMEM_am_dmessage_t *) MPL_malloc(sizeof(MPIDI_XPMEM_am_dmessage_t),
+                                                     MPL_MEM_OTHER);
+        MPIR_ERR_CHKANDJUMP(dmessage == NULL, mpi_errno, MPI_ERR_NO_MEM, "**nomem");
+
+        MPIR_Memcpy(&dmessage->ctrl_hdr, ctrl_hdr, sizeof(MPIDI_SHM_ctrl_hdr_t));
+        DL_APPEND(MPIDI_XPMEM_global.dmessage_queue, dmessage);
+    } else {
+        mpi_errno = MPIDI_XPMEM_ctrl_send_lmt_req_cb(ctrl_hdr);
+        MPIR_ERR_CHECK(mpi_errno);
+    }
+
+  fn_exit:
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+int MPIDI_XPMEM_ctrl_send_lmt_cts_req_cb(MPIDI_SHM_ctrl_hdr_t * ctrl_hdr)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_Request *sreq;
+    MPIDI_SHM_ctrl_xpmem_send_lmt_req_t *slmt_req_hdr = &ctrl_hdr->xpmem_slmt_req;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_XPMEM_CTRL_SEND_LMT_CTS_REQ_CB);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_CTRL_SEND_LMT_CTS_REQ_CB);
+
+    /* Sender gets the CTS packet from receiver and need to perform copy */
+    sreq = (MPIR_Request *) slmt_req_hdr->sreq_ptr;
+    MPIDIG_REQUEST(sreq, rank) = slmt_req_hdr->src_rank;
+    MPIDIG_REQUEST(sreq, tag) = slmt_req_hdr->tag;
+    MPIDIG_REQUEST(sreq, context_id) = slmt_req_hdr->context_id;
+
+    /* Complete XPMEM receive */
+    mpi_errno =
+        MPIDI_XPMEM_handle_lmt_recv(slmt_req_hdr->src_offset, slmt_req_hdr->data_sz,
+                                    slmt_req_hdr->rreq_ptr, slmt_req_hdr->src_lrank,
+                                    MPIDI_SHM_XPMEM_SEND_LMT_CTS, sreq);
```

Comment:
- it calls `MPIDI_XPMEM_handle_lmt_recv` because sender also performs receiving operation like receiver after getting CTS packet for cooperative copy, so here I still let it call `MPIDI_XPMEM_handle_lmt_recv`. Inside `MPIDI_XPMEM_handle_lmt_recv`, I will distinguish sender and receiver, and make them process data accordingly.

- you mean I split `MPIDI_XPMEM_handle_lmt_recv` into two subroutines for sender and receiver?

Reply:
Yes, I mean define two routines, one for sender side and the other for receiver side. Otherwise, the function naming is very confusing.